src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+        IEnumerable<SyntaxNode> descendantNodes = constructor.DescendantNodes();
+
+        HandleArgumentsList(descendantNodes, context, fields, checkConstructorParameters: arguments.Count != 0);
+        HandleInvocations(descendantNodes, context, fields);
+    }
+
+    /// <summary>
+    /// This method handles assignment expression in constructors to find field initializer candidates.
+    /// </summary>
+    /// <param name="descendantNodes">The constructor's descendant nodes.</param>
+    /// <param name="context">The node analysis context.</param>
+    /// <param name="fields">A dictionary tracking all fields intializations in constructors.</param>
+    /// <param name="checkConstructorParameters">A bool on whether to check constructor parameters for field initializations.</param>
+    private static void HandleArgumentsList(
+        IEnumerable<SyntaxNode> descendantNodes,
+        SyntaxNodeAnalysisContext context,
+        Dictionary<string, FieldInitializationInfo> fields,
+        bool checkConstructorParameters)
+    {
+        foreach (AssignmentExpressionSyntax assignment in descendantNodes.OfType<AssignmentExpressionSyntax>())
+        {
+            bool isFieldAssignment = assignment.Left is IdentifierNameSyntax;
+
+            // We want to ignore assignments that are not part of an expression statement or are not an identifier
+            if (assignment.Parent is not ExpressionStatementSyntax
+                || !(assignment.Left is MemberAccessExpressionSyntax
+                || isFieldAssignment))
+            {
+                continue;
+            }
+
+            IdentifierNameSyntax identifierName = isFieldAssignment ? (IdentifierNameSyntax)assignment.Left : (IdentifierNameSyntax)((MemberAccessExpressionSyntax)assignment.Left).Name;
+
+            // We assume the identifier is a field. Calls to semantic model are super expensive, so we want to avoid them as much as possible.
+            string fieldName = identifierName.Identifier.Text;
+            if (!fields.TryGetValue(fieldName, out FieldInitializationInfo fieldInfo))
+            {
+                fieldInfo = new FieldInitializationInfo(fieldName);
+                fields.Add(fieldName, fieldInfo);
+            }
+
+            if (fieldInfo.ShouldNotInitializeInDeclaration)
+            {
+                continue;
+            }
+
+            if (assignment.Right is InvocationExpressionSyntax invocationExpressionSyntax
+                && !IsStaticMethodInvocation(invocationExpressionSyntax, context))
+            {
+                // If the assignment is to a non-static method invocation, we can ignore it.
+                fieldInfo.ShouldNotInitializeInDeclaration = true;
+                continue;
+            }
+
+            if (checkConstructorParameters)
+            {
+                IOperation? operation = context.SemanticModel.GetOperation(assignment.Right, context.CancellationToken);
```

Comment:
Note: This is expensive

Reply:
I got rid of this entirely in my refactor